### PR TITLE
Pairing gadget tidying (depends on #39)

### PR DESCRIPTION
**Pull request by dtebbs**
Some interface updates, comments and functionality related to pairing gadgets (particularly for BW6-761 / BLS12-377).

### CODE REVIEW

File: libsnark/gadgetlib1/gadgets/curves/weierstrass_g1_gadget.hpp [hunk 1]
```diff
@@ -43,6 +41,17 @@ template<typename ppT> class G1_variable : public gadget<libff::Fr<ppT>>
         const libff::G1<other_curve<ppT>> &P,
         const std::string &annotation_prefix);
 
+    // NOTE: pb and annotation_prefix are redundant here, but required because
+    // of the inheritance from gadget (and in order to construct a new
+    // pb_linear_combination in operator-()).
+    G1_variable(
```
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/curves/weierstrass_g1_gadget.hpp:47]: Related to https://github.com/clearmatics/libsnark/issues/36

File: libsnark/gadgetlib1/gadgets/pairing/bw6_761_bls12_377/bls12_377_miller_loop.hpp [hunk 2]
```diff
@@ -82,6 +82,35 @@ class bls12_377_miller_loop_gadget : public gadget<libff::Fr<ppT>>
     void generate_r1cs_witness();
 };
 
+template<typename ppT>
+class bls12_377_e_over_e_miller_loop_gadget : public gadget<libff::Fr<ppT>>
```
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/pairing/bw6_761_bls12_377/bls12_377_miller_loop.hpp:86]: This PR partially addresses #33

File: libsnark/gadgetlib1/tests/test_pairing.cpp [hunk 3]
```diff
@@ -140,58 +139,83 @@ template<typename ppT> void test_mnt_miller_loop(const std::string &annotation)
...
+    test_miller_loop<ppT>(P_val, Q_val, native_result, annotation);
+}
+
+template<typename ppT>
+void test_e_over_e_miller_loop(
+    const libff::G1<other_curve<ppT>> &P1_val,
+    const libff::G2<other_curve<ppT>> &Q1_val,
+    const libff::G1<other_curve<ppT>> &P2_val,
+    const libff::G2<other_curve<ppT>> &Q2_val,
+    const FqkT<ppT> &expect_result,
```
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/tests/test_pairing.cpp:193]: This is called `native_result` in the `test_miller_loop` function above. Let's use consistent names to represent the expected result here.

**Review by AntoineRondelet — COMMENTED** [review]
Few little comments and then let's rebase. Quick one not strictly related to this PR but mentioning it here: https://github.com/clearmatics/libsnark/pull/40/files#diff-9c8bfda3b281caab60a32791eb34805afe5dfe7897ef8b1b49cb900187605df2R410 (excerpt below)
```
        // f <- f^2 * ell_Q1(P1)
        _f_ell_P.emplace_back(new bls12_377_ate_compute_f_ell_P<ppT>(
            pb,
            *P4_prec._Px,
            _minus_P4_Y,
            *Q4_prec._coeffs[coeff_idx],
            *f,
            Fp12_2over3over2_variable<FqkT>(
                pb, FMT(annotation_prefix, " f^2*ell_Q4(P4)")),
            FMT(annotation_prefix, " _f_ell_P[%zu]", _f_ell_P.size())));
f = &_f_ell_P.back()->result();
```

-> this comment needs to be fixed I believe, here we're not accumulating evals at P1 but should instead of P4 afaict.

File: libsnark/gadgetlib1/tests/test_pairing.cpp (hunk 3 above)
Comment (dtebbs) [reply libsnark/gadgetlib1/tests/test_pairing.cpp:193]: Renamed to `expect_result`.

Comment (AntoineRondelet) [reply libsnark/gadgetlib1/tests/test_pairing.cpp:193]: Thanks

**Review by AntoineRondelet — APPROVED** [review]
LGTM

**Pull request merged**